server/src/main/java/org/elasticsearch/index/mapper/MapperMergeValidator.java

```diff
@@ -0,0 +1,184 @@
+     * of a field alias, corresponds to a valid part of the mapping.
+     */
+    public static void validateFieldReferences(List<FieldMapper> fieldMappers,
+                                               List<FieldAliasMapper> fieldAliasMappers,
+                                               Map<String, ObjectMapper> fullPathObjectMappers,
+                                               FieldTypeLookup fieldTypes) {
+        validateCopyTo(fieldMappers, fullPathObjectMappers, fieldTypes);
+        validateFieldAliasTargets(fieldAliasMappers, fullPathObjectMappers);
+    }
+
+    private static void validateCopyTo(List<FieldMapper> fieldMappers,
+                                       Map<String, ObjectMapper> fullPathObjectMappers,
+                                       FieldTypeLookup fieldTypes) {
+        for (FieldMapper mapper : fieldMappers) {
+            if (mapper.copyTo() != null && mapper.copyTo().copyToFields().isEmpty() == false) {
+                String sourceParent = parentObject(mapper.name());
+                if (sourceParent != null && fieldTypes.get(sourceParent) != null) {
+                    throw new IllegalArgumentException("[copy_to] may not be used to copy from a multi-field: [" + mapper.name() + "]");
+                }
+
+                final String sourceScope = getNestedScope(mapper.name(), fullPathObjectMappers);
+                for (String copyTo : mapper.copyTo().copyToFields()) {
+                    String copyToParent = parentObject(copyTo);
+                    if (copyToParent != null && fieldTypes.get(copyToParent) != null) {
+                        throw new IllegalArgumentException("[copy_to] may not be used to copy to a multi-field: [" + copyTo + "]");
+                    }
+
+                    if (fullPathObjectMappers.containsKey(copyTo)) {
+                        throw new IllegalArgumentException("Cannot copy to field [" + copyTo + "] since it is mapped as an object");
+                    }
+
+                    final String targetScope = getNestedScope(copyTo, fullPathObjectMappers);
+                    checkNestedScopeCompatibility(sourceScope, targetScope);
+                }
+            }
+        }
+    }
+
+    private static void validateFieldAliasTargets(List<FieldAliasMapper> fieldAliasMappers,
+                                                  Map<String, ObjectMapper> fullPathObjectMappers) {
+        for (FieldAliasMapper mapper : fieldAliasMappers) {
+            String aliasName = mapper.name();
+            String path = mapper.path();
+
+            String aliasScope = getNestedScope(aliasName, fullPathObjectMappers);
+            String pathScope = getNestedScope(path, fullPathObjectMappers);
+            if (!Objects.equals(aliasScope, pathScope)) {
+                throw new IllegalArgumentException("Invalid [path] value [" + path + "] for field alias [" +
+                    aliasName + "]: an alias must have the same nested scope as its target.");
```

Comment:
maybe add what we computed as the scope of the alias and the path to the error message

Reply:
👍